Add ledger configuration node-to-client query

DavidEichmann opened the pull request:
Key changes:

* Add `GetLedgerConfig :: Query blk (LedgerConfig blk)`
* Make `SerialiseNodeToClient blk (LedgerConfig blk)` a superclass of `SerialiseNodeToClientConstraints`
  * This makes sense now that the Node-to-Client protocol includes a `GetLedgerConfig` query.
* Add CBOR serialization instances for ledger configs and nested types.

DavidEichmann commented:
I've pushed an update but still need to do another pass to clean things up, tomorrow. This update primarily changes from a `GetPartialLedgerConfig` query to a `GetLedgerConfig` quert. Additionally, it rebase's on duncan's recent changes that adds a new top level query for the system start time.

coot commented:
I removed myself, @bolt12 and @karknu from the reviewers; I don't see anything related to the networking (am I right?, the scope of this PR is quite big).

DavidEichmann commented:
@coot That's fine. I'm going over a few points then I assume @nfrisby will review.

DavidEichmann commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +207,35 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+      triggerHardFork
+    )
+      = encodeListLen 2
+        <> encodeShelleyLedgerConfigWith mempty ccfg version shelleyLedgerConfig
```
list length used in `encodeShelleyLedgerConfigWith` will be wrong since we use `mempty` here

DavidEichmann commented:
This is now blocked on https://github.com/input-output-hk/cardano-base/pull/222

DavidEichmann commented:
@nfrisby I'm waiting for review here.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Simple.hs:
```diff
@@ -20,3 +27,17 @@ data TriggerHardFork =
   | TriggerHardForkNever
   deriving (Show, Generic, NoThunks)
+
+instance SerialiseNodeToClient blk TriggerHardFork where
```
My initial thought is that this would also belong in the `Ouroboros.Consensus.HardFork.Combinator.Serialisation.SerialiseNodeToClient` module.

But if I recall correctly, I think you had asked why that module of orphans exists. I don't have a good answer (yet?).

DavidEichmann replied:
Orphan instances are a consistent source of annoyance to me. I'd prefer to avoid them where possible.

DavidEichmann replied:
I notice often e.g. [here](https://github.com/input-output-hk/ouroboros-network/pull/3214#discussion_r668905769), that existing orphans encourage new orphans.

Here is my explanation of the mechanism. We have an "orphans" module, `O`, that imports many modules/types and hence will be near the top of the module dependency tree. When adding a new instance in module `M` that requires one of those orphan instances, it's likely that importing `O` from `A` will cause a module cycle. This cycle could be resolved by moving the relevant instances out of `O` to their respective types (i.e. making them non-orphan instances), but the instances often depend on each other and hence would require even more instance be moved out of `O`. We often just pick the easy option and move our new instance from `M` to `O` as an orphan instance. Perhaps there was some good reason to start this off in the first place, but I wouldn't be surprised if most instances are orphans do to this mechanism.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:
```diff
@@ -195,6 +221,7 @@ answerQuery ::
 answerQuery cfg query st = case query of
   BlockQuery blockQuery -> answerBlockQuery cfg blockQuery st
```
Style: all these `->`s should be aligned.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/Counting.hs:
```diff
@@ -66,6 +72,16 @@ import           Ouroboros.Consensus.Util.SOP
 
 newtype Exactly xs a = Exactly { getExactly :: NP (K a) xs }
 
+instance (SerialiseNodeToClient blk a, SListI xs)
```
Optional thought: This logic seems like the write logic for serializing any `NP`. Do we already have an instance for that? `... => ToCBOR (NP f xs)` and `... => FromCBOR (NP f xs)` eg?

DavidEichmann replied:
I think this implementation is actually wrong regarding versioning. It should omit elements for disabled eras as is done in `SerialiseNodeToClient (HardForkBlock xs) (PerEraLedgerConfig xs)`. I'll try to generalize an encoder/decoder for `NP f xs`.

DavidEichmann replied:
Oh my! That turned out to be a bit of a headache, but I think I managed. In the end I had to:

* Move `SerialiseNodeToClient (HardForkBlock xs) (HardForkLedgerConfig xs)` from `Ouroboros.Consensus.HardFork.Combinator.Basics` to `Ouroboros.Consensus.HardFork.Combinator.Serialisation.SerialiseNodeToClient` (I still dislike orphan instances, but with all the module cycle that I had to deal with, I finally gave in :shrug: ).
* Added a generalized encoding/decoding for `NP f xs`, namely `encodeNodeToClientNP` and `encodeNodeToClientNP` and use it to implemented `SerialiseNodeToClient` instances for `Shape xs` and `PerEraLedgerConfig xs`

I'll push these change after I respond to the remaining comments.

nfrisby commented on ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Node.hs:
```diff
@@ -284,4 +287,5 @@ instance NodeInitStorage ByronBlock where
   isSelfIssued = isSelfIssuedConstUnknown
 
-instance RunNode ByronBlock
+instance SerialiseNodeToClient ByronBlock (LedgerConfig ByronBlock)
```
Doesn't the `SerialiseNodeToClient ByronBlock Cardano.Chain.Genesis.Config` instance in the ` Ouroboros.Consensus.Byron.Node.Serialisation` module provide this? Can we drop this new instance context element?

DavidEichmann replied:
Yes! Will delete

nfrisby commented on ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Node/Serialisation.hs:
```diff
@@ -148,7 +152,12 @@ instance SerialiseNodeToNode ByronBlock (GenTxId ByronBlock) where
 -------------------------------------------------------------------------------}
 
-instance SerialiseNodeToClientConstraints ByronBlock
+instance SerialiseNodeToClient ByronBlock (LedgerConfig ByronBlock)
```
Doesn't the `SerialiseNodeToClient ByronBlock Cardano.Chain.Genesis.Config` instance below provide this? Can we drop this new instance context element?

DavidEichmann replied:
yup

nfrisby commented on ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs:
```diff
@@ -80,9 +80,12 @@ import           Cardano.Binary (ToCBOR (..))
 import           Cardano.Crypto.Hash (Hash, HashAlgorithm, MD5, ShortHash)
 import qualified Cardano.Crypto.Hash as Hash
 
+import           Ouroboros.Network.Block
```
I generally don't expect to see this import added anywhere. Do you know what exactly its bringing in scope here? I anticipate there is a `Consensus` module that re-exports that thing, whatever it is.

DavidEichmann replied:
It brings `Serialised` (defined directly in `Ouroboros.Network.Block`)  into scope which is used [below](https://github.com/input-output-hk/ouroboros-network/pull/3214/files#diff-b3aa9feded0c37a4ee5a60931752113f71b8f12a2a109c6645198b9cca26c9ceR142) for the `SerialiseNodeToClientConstraints (SimpleBlock c ext)` instance. So this is not something that can be trivially changed.

nfrisby replied:
Yep, looks like no `Consensus` module re-exports that one :(

Would you write an explicit import list on that one? It will avoid me wondering this same question at a future date :)

nfrisby commented on ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Node.hs:
```diff
@@ -61,6 +63,7 @@ instance ( LedgerSupportsProtocol      (SimpleBlock SimpleMockCrypto ext)
          , Show (ForgeStateInfo        (SimpleBlock SimpleMockCrypto ext))
          , Show (ForgeStateUpdateError (SimpleBlock SimpleMockCrypto ext))
          , Serialise ext
+         , SerialiseNodeToClient (SimpleBlock SimpleMockCrypto ext) (LedgerConfig (SimpleBlock SimpleMockCrypto ext))
```
This is instance is declared in Ouroboros.Consensus.Mock.Ledger.Block`, no?

DavidEichmann replied:
Indeed, I can replace this with `Serialise (MockLedgerConfig SimpleMockCrypto ext)`

nfrisby commented on ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs:
```diff
@@ -337,9 +354,16 @@ data SimpleLedgerConfig c ext = SimpleLedgerConfig {
 
+instance MockProtocolSpecific c ext => HasPartialLedgerConfig (SimpleBlock c ext)
+
+instance (Serialise (MockLedgerConfig c ext))
```
Oof. It's a bummer that the `SimpleLedgerConfig` includes a `MockLedgerConfig` type family whereas the `MockBlock` is a synonym for a specialization of `SimpleBlock`; that naming seems upside-down :(

DavidEichmann replied:
Yes.. It's a bit odd. Can we address this as a separate issue?

nfrisby replied:
Right, yes. I was just sharing a grumble/sympathy, not requesting any action.

nfrisby commented on ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Block.hs:
```diff
@@ -130,6 +135,18 @@ instance (SimpleCrypto c, Serialise ext') => Serialise (SimpleBlock' c ext ext')
+          , SerialiseNodeToClient (SimpleBlock c ext) (SomeSecond BlockQuery (SimpleBlock c ext))
+          , SerialiseNodeToClient (SimpleBlock c ext) (LedgerConfig (SimpleBlock c ext))
+          , SerialiseResult       (SimpleBlock c ext) (BlockQuery (SimpleBlock c ext))
+          ) => SerialiseNodeToClientConstraints (SimpleBlock c ext)
```
Did you need to generalize the `MockBlock` instance, or did it just seem like the right thing to do? I wouldn't be surprised if we only ever required `RunNode` etc for the `MockBlock` specialization of `SimpleBlock`, FWIW.

DavidEichmann replied:
Yeah, I didn't need to generalize. I was trying to avoid an orphan instance, but I've reverted this so as to keep the diff cleaner.

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -78,6 +78,7 @@ fromShelleyLedgerExamples ShelleyLedgerExamples {
     , exampleQuery            = queries
     , exampleResult           = results
     , exampleAnnTip           = unlabelled annTip
+    , exampleLedgerConfig     = mempty
```
See similar comment in the Cardano `Examples` module.

DavidEichmann replied:
I can easily stick a `ShelleyLedgerConfig era` in here, but we'd end up with the same example for each era. Perhaps I can use something in the `ShelleyLedgerExamples` parameter as a source of noise. Did you have something else in mind?

DavidEichmann replied:
This will be replaced with some dummy example

nfrisby replied:
Fixed! Marking Resolved.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:
```diff
@@ -104,8 +104,18 @@ instance ShelleyBasedEra era => NoThunks (ShelleyLedgerError era)
+      -- when the top-level block is a 'HardForkBlock' that contains
+      -- 'ShelleyBlock' -- such a 'ShelleyLedgerConfig' will have been
+      -- /completed/ by the HFC to contain the most information 'EpochInfo'
+      -- possible base on the past hard forks (and possibly also the next hard
```
`base on` -> `based on`

Nice comment 👍

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:
```diff
@@ -405,7 +406,9 @@ instance ShelleyBasedEra era => BlockSupportsMetrics (ShelleyBlock era) where
          NotSelfIssued -> IsNotSelfIssued
 
-instance ShelleyBasedEra era => RunNode (ShelleyBlock era)
+instance ( SerialiseNodeToClientConstraints (ShelleyBlock era)
```
My initial expectation is that we shouldn't need to add `SerialiseNodeToClientConstraints` here.

Edit: Hmm. seems like you're using this to workaround module hierarchies/avoiding orphans/etc. Is that it? I'm not ecstatic about this approach, but I don't yet obviously see simple path to something cleaner.

DavidEichmann replied:
Yes, this is something I've struggled with a few times: I add some "unneeded" constraints because importing the relevant orphan instance would cause a module cycle. This is one of the reasons to dislike orphan instances. I also don't see an clear solution short of getting rid of the relevant orphan instances which sounds like a significant refactoring though.

Trying to move this instance into `Ouroboros.Consensus.Shelley.Ledger.Block` so that it itself is at least not an orphan is messy too. We require most of the other orphan instances in this module and can't import it without creating another module cycle. The existing orphans in a way force us to make more orphans :unamused: .

nfrisby replied:
Would you open a GitHub Issue in this repo, labeled `consensus` and `tech debt` about trying to wrangle these orphans? Or at least better document the method to the madness?

I agree it's likely a significant refactoring, but I think the state of the problem is better represented in your head right now than it has been for anyone, maybe ever. If you could write some of that insight down it could help us eventually approach that task 🙏

DavidEichmann replied:
Done #3302

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node/Serialisation.hs:
```diff
@@ -9,9 +15,10 @@ import           Control.Exception (Exception, throw)
+import           Cardano.Binary
 import           Codec.Serialise (decode, encode)
 
+
```
Style: please don't introduce this newline.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -189,6 +201,11 @@ data ShelleyPartialLedgerConfig era = ShelleyPartialLedgerConfig {
       -- some rather expensive computations that shouldn't be repeated too
       -- often (e.g., 'sgActiveSlotCoeff').
+      --
+      -- TODO Specialize this to a ShelleyLedgerConfig with unit instead of
```
So do we need to convince Clarke of the necessity of input-output-hk/cardano-ledger-specs#2344 after all? My last comment there is saying it's merely a nice-to-have at this point. Is that still accurate?

DavidEichmann replied:
This is just a nice-to-have. I'll remove the comment for now.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +215,126 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+        <> toCBOR @Natural maxMajorPV'
+        <> toCBOR @Word64 maxLovelaceSupply'
+
+        -- TODO shouldn't all these things use `SerialiseNodeToClient` instances?
```
Reifying this TODO as a PR comment.

DavidEichmann replied:
These are all things defined in lower level packages. I don't see a benefit to making explicit `SerialiseNodeToClient` for these types, so I'll leave it like this for now and remove the comment.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +215,126 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+
+-- | This instance uses the invariant that the EpochInfo in a
+-- ShelleyPartialLedgerConfig is always just a dummy value.
+instance ShelleyBasedEra era
```
`decode` expects 13 and `encode` expects 15 but I count 14 😆

DavidEichmann replied:
Wow. yes. But how could this have passed the golden tests?! Those should check that the CBOR is valid.

DavidEichmann replied:
Ok, So I'm looking at `Test.Util.Serialisation.Golden`'s `goldenTestCBOR`. It seems that we do some amount of checking for valid CBOR, but this is not extensive.

In the case of the golden file, we only check that we're able to decode a series of `TermToken`s (see `decodeAsFlatTerm`). We check the individual tokens are valid, but not that the sequence of tokens are valid as a whole. Things like incorrect list length are not detected (hence why this passed the golden tests)!

In the case of the actual value, we only check that the encoding doesn't throw an exception (though this is still a passed test if the golden file contains the same exception).

Note that we never actually test `decodeNodeToClient`, so the `enforceSize` call never gets a chance to check the list length either. Yes, you heard me right! Our golden tests only test the encode, not the decoder!

What I propose we do here is:
* [ ] I fix this instance with the correct list length of 14 (yes I double checked)
* [ ] In a separate PR I change goldenTestCBOR to...
  * ... use [validFlatTerm](https://hackage.haskell.org/package/cborg-0.2.5.0/docs/Codec-CBOR-FlatTerm.html#v:validFlatTerm) to check that both the "golden" decoding and the "actual" encoding are valid terms. I hope there are no existing golden tests with invalid CBOR :grimacing:.
  * ... decode the golden file using  `decodeNodeToClient` to test the decoding half of our serialization code.

nfrisby replied:
We have separate Roundtrip tests; are we just missing those for this type maybe?

DavidEichmann replied:
I'd say you're right, I should add round trip tests, but also it seems odd to me that we only test serializes and not deserializers in the golden tests. We really do want to test the the deserializers will succeed on the golden files.

nfrisby replied:
I have a slight preference for not testing the golden files deserialize -- leaving the Round-Trip Tests to find bugs there. ... but your idea would catch an "out-of-date" golden test file: suppose an invariant is added to the deserializer (and the generators) but we forget to update the golden files. Let's: compute is cheap here 👍

DavidEichmann replied:
So I think in terms of this PR, the tests are sufficient. One might say "we're golden" :laughing:.

My concern with not testing deserialization of golden files was that our deserializers aren't tested on old node-to-client versions. You're saying the golden tests are sufficient. Indeed, the golden tests run across all supported node-to-client versions. My thought was then: but what if we change our serializers *and* deserializers in a compatible way for an old node-to-clinet version? The round trip tests will pass but we're not really testing against the correct serialization. But indeed, we'd have to change the serialization code in this scenario so a gold test that only tests serialization would catch that regression. So perhaps you're right! I'm happy to leave this as is for now.

nfrisby replied:
Sounds good, thanks. Resolving.

nfrisby commented on ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator/A.hs:
```diff
@@ -199,8 +198,24 @@ data PartialLedgerConfigA = LCfgA {
   deriving NoThunks via OnlyCheckWhnfNamed "LCfgA" PartialLedgerConfigA
 
-type instance LedgerCfg (LedgerState BlockA) =
-    (EpochInfo Identity, PartialLedgerConfigA)
```
Was this `EpochInfo Identity` just dead-code?

DavidEichmann replied:
Yes, as far as I can see.

nfrisby replied:
That's a good observation. But I think we should remove the dead-code.

If we want to emphasize the `EpochInfo` thing here, then we should figure out a way to make it not dead -- ie to actually need it.

DavidEichmann replied:
I'm leaning toward keeping the current change (i.e. removing the `EpochInfo` from the partial ledger config). In the case of `ByronBlock` we also don't add an `EpochInfo` in it's partial ledger config, hence there is precedent.

DavidEichmann replied:
> But I think we should remove the dead-code.

I'm reading this as "we should remove the `EpochInfo` from `LedgerCfg (LedgerState BlockA)`". This is in fact what I've done. Is there something else you want changed here?

nfrisby replied:
Yeah; maybe the code wasn't as fixed-up when I originally wrote that? Or maybe I was confused.

It looks good now. Resolving. Thanks.

nfrisby commented on ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator/A.hs:
```diff
@@ -199,8 +198,24 @@ data PartialLedgerConfigA = LCfgA {
+  encodeNodeToClient _ _ = toCBOR
+  decodeNodeToClient _ _ = fromCBOR
+
+type instance LedgerCfg (LedgerState BlockA) = PartialLedgerConfigA
```
This naming is incongruous. We should probably rename `PartialLedgerConfigA` to `LedgerConfigA`

nfrisby commented on ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator/B.hs:
```diff
@@ -418,6 +432,7 @@ instance Serialise (SerialisedHeader BlockB) where
   SerialiseNodeToClient
 -------------------------------------------------------------------------------}
 
+instance SerialiseNodeToClient BlockB ()
```
~~Why is this needed? Serialization for `()` is suspect; it's usually something we'll just omit from the serialization instead of using `TkNull`. And that means there's a `toCBOR` or `fromCBOR` that should be counted against the list length, etc etc.~~

Ah, it's because `type instance LedgerCfg (LedgerState BlockB) = ()` -- no surprise. This is probably fine for such a testing module. But it'd be "safer" to declare `data LedgerCfgB = LedgerCfgB` and only provide instances for that :/

nfrisby commented on ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs:
```diff
@@ -121,6 +143,7 @@ instance Inject Examples where
       , exampleQuery            = inj (Proxy @(SomeSecond BlockQuery)) exampleQuery
       , exampleResult           = inj (Proxy @SomeResult)              exampleResult
       , exampleAnnTip           = inj (Proxy @AnnTip)                  exampleAnnTip
+      , exampleLedgerConfig     = mempty -- We rely on `combineEras` to make use of the example ledger configs
```
My initial expectation is that this field should be a use of `inj` just like all the others. I can imagine that it was not obvious how to define the different examples on a per-era basis for Shelley, though. I have a couple ideas for that, if you confirm that's the only reason you went with `mempty` for now.

DavidEichmann replied:
Yes, indeed we're trying to inject the ledger config for a single era but we need a ledger config for all eras to produce an example here. Happy to hear your ideas.

DavidEichmann replied:
So I think I've found a simpler solution. I think we were implicitly insisting on keeping the `Inject Examples` instance, when really we can just replace it's use with a `mkCardanoExamples :: NP Examples (CardanoEras Crypto) -> Examples (CardanoBlock Crypto)` which is able to use `inject` as usual for other fields but also combine the per-era ledger configs into a HFC ledger config. This solution feels correct and like a simplification to the PR.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -135,6 +139,50 @@ dispatchDecoderErr ccfg version =
+      x :* _ -> case ccfgs of
+        cfg :* _ -> encodeNodeToClient cfg versionX x
+    HardForkNodeToClientEnabled _ subVersions ->
+      let -- Encoding of the components of xs, filtering out the eras disabled in subVersions.
```
Are we filtering because if we negotiated a version which excludes later eras that we're aware of that means they are not even aware of those versions, and so they wouldn't expect them in the serialization?

That makes sense, but it would probably be worth adding a comment on this instance that spells that out for future clarity 👍

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -135,6 +139,50 @@ dispatchDecoderErr ccfg version =
+instance All SerialiseConstraintsHFC xs
+  => SerialiseNodeToClient (HardForkBlock xs) (PerEraLedgerConfig xs) where
+  encodeNodeToClient (HardForkCodecConfig (PerEraCodecConfig ccfgs)) version (PerEraLedgerConfig xs) = case version of
+    HardForkNodeToClientDisabled versionX -> case xs of
```
Optional: consider if `case (xs, ccfgs) of` would be easier to read.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -135,6 +139,50 @@ dispatchDecoderErr ccfg version =
+            $ hczipWith3
+                (Proxy @SerialiseConstraintsHFC)
+                (\ccfg subVersionMay x -> K $ case subVersionMay of
+                  EraNodeToClientEnabled subVersion -> Just (encodeNodeToClient ccfg subVersion x)
```
Style: align the `->`

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -135,6 +139,50 @@ dispatchDecoderErr ccfg version =
+    ledgerConfigs <- hsequence' $ hczipWith
+      (Proxy @SerialiseConstraintsHFC)
+      (\ccfg subVersionMay -> Comp $ case subVersionMay of
+          EraNodeToClientDisabled -> failVersion
```
Style: align the `->`

nfrisby reviewed (changes requested):
Phew! That's required some focus :D. Thanks for making it so clean.

I made several comments throughout, ranging from style to suspicions of leftover comments/minor alterations, to some higher level requests/suggestions/concerns.

DavidEichmann commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Summary.hs:
```diff
@@ -64,10 +66,10 @@ import           Cardano.Binary (enforceSize)
-
 import           Ouroboros.Consensus.HardFork.History.EraParams
 import           Ouroboros.Consensus.HardFork.History.Util
+import           Ouroboros.Consensus.Node.Serialisation
```
Will fix.

Is there some secret to knowing which package each import is from? I find my self doing a lot of text search to try and figure out where to put each new import statement.

nfrisby replied:
See bullet point 15 "Import lists" in https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/docs/StyleGuide.md

That's the most indicative explanation. But eg it doesn't mention the HardFork thing; I'm only aware of that because of having seen it set apart in the existing code.

DavidEichmann commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/PartialConfig.hs:
```diff
@@ -50,8 +54,7 @@ class ( ConsensusProtocol p
   completeConsensusConfig _ _ = id
 
 -- | Partial ledger config
-class ( UpdateLedger blk
```
I think it was causing issue at some point, but now `cabal test all` succeeds so I'm not sure what happened there. I'll revert this change.

nfrisby replied:
Please leave the formatting the way it was. Less noise in the diff, etc.

DavidEichmann commented on ouroboros-consensus-byronspec/src/Ouroboros/Consensus/ByronSpec/Ledger/Genesis.hs:
```diff
@@ -55,6 +67,31 @@ data ByronSpecGenesis = ByronSpecGenesis {
+      <*> fromCBOR @Spec.BlockCount
+      <*> fromCBOR @Natural
+
+-- TODO These instances should be moved upstream
```
https://github.com/input-output-hk/cardano-ledger-specs/pull/2378

DavidEichmann commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/Common.hs:
```diff
@@ -229,6 +231,7 @@ class ( SingleEraBlock                   blk
       , SerialiseDiskConstraints         blk
       , SerialiseNodeToNodeConstraints   blk
       , SerialiseNodeToClientConstraints blk
+      , SerialiseNodeToClient            blk (PartialLedgerConfig blk)
```
`SerialiseNodeToClientConstraints blk` already implies a `SerialiseNodeToClient blk (LedgerConfig blk)` constraint. The constraint I'm adding here is for the *partial* ledger config. This is HFC specific as the `HardForkLedgerConfig` contains the partial ledger configs of the individual blocks. So I think this is in the correct place. Does that make sense?

DavidEichmann commented:
@nfrisby I've pushed some changes in response to your previous round of comments (most of which I've resolved). This includes adding round trip tests for ledger config serialization. The round trip tests which I ran locally to confirm that it fails due to an incorrect list length that you pointed out. Then, I fixed the encoding and got the tests to pass (you know.... test driven development style!).
I separately pushed a rebase on the latest master.

DavidEichmann commented:
Some minor updates:

* Removed "WIP" from the latest commit comment on adding round trip tests.
* Removed unused import
* 
Since your last review @nfrisby, I suggest you look at these:

* The remaining unresolved PR comments
* [This diff](https://github.com/input-output-hk/ouroboros-network/compare/27112663d722b36060de4553389536a425f1ff11..4680fc691c337a6d9530ceb03353c5a488879189) responding to the now resolved PR comments
* the latest commit which adds round trip tests

DavidEichmann commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -135,6 +261,16 @@ dispatchDecoderErr ccfg version =
+  encodeNodeToClient ccfg version (PerEraLedgerConfig xs)
+    = encodeNodeToClientNP encodeNodeToClient ccfg version xs
+  decodeNodeToClient ccfg version
+    = PerEraLedgerConfig <$> decodeNodeToClientNP decodeNodeToClient ccfg version
```
@nfrisby You mentioned that we have an invariant that using `blk ~ A` and using `blk ~ HardForkBlock '[A]` should be compatible. This is sort of broken: the encoding for `LedgerConfig A` and `LedgerConfig HardForkBlock '[A]` will be different as they are different types. We'd have to make an attempt to convert the `LedgerConfig HardForkBlock '[A]` back into a `LedgerConfig A`. Does that seem like the correct thing to do?

nfrisby replied:
I think they only have to be "(morally) isomorphic", not necessarily "compatible". IE I'm not sure we would need an `A` node and an `HFC '[A]` node to actually be able to communicate.

I'm guessing there are already other `SerialiseNodeTo*` instances where the `HFC` layer leads to incompatibilities. Please check me on that.

DavidEichmann replied:
Humm yeah. I think you're right here. We do want to be backwards compatible with old clients that use `A` instead of `HFB '[A]`, but such old clients will not know of the new `GetLedgerConfig` query and hence will never request/decode it. So we're actually still compatible in that sense.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +210,124 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+            )
+        <*> fromCBOR @(SL.TranslationContext era)
+      )
+      <*> decodeNodeToClient @(ShelleyBlock era) @TriggerHardFork  ccfg version
```
There's an extra space between `@TriggerHardFork` and `ccfg`

nfrisby commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs:
```diff
@@ -106,6 +108,37 @@ roundtrip' enc dec a = case deserialiseFromBytes dec bs of
+  -> Property
+roundtripComparingEncoding enc dec = roundtripComparingEncoding' enc (const <$> dec)
+
+-- | Like roundtrip', but checks for equality on of the encoding (i.e. the byte
```
typo: `on of`

nfrisby commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs:
```diff
@@ -370,6 +407,7 @@ roundtrip_SerialiseNodeToClient
      , ArbitraryWithVersion (BlockNodeToClientVersion blk) (ApplyTxErr blk)
      , ArbitraryWithVersion (BlockNodeToClientVersion blk) (SomeSecond BlockQuery blk)
      , ArbitraryWithVersion (BlockNodeToClientVersion blk) (SomeResult blk)
+     , Arbitrary (WithVersion (BlockNodeToClientVersion blk) (LedgerConfig blk))
```
This juxtaposition lead to a surprise that is a real bummer, but I can fully understand how it came to be.

Would you open a ticket before we merge this PR to rename `ArbitraryWithVersion`?

DavidEichmann replied:
https://jira.iohk.io/browse/CAD-3293

nfrisby commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs:
```diff
@@ -383,6 +421,15 @@ roundtrip_SerialiseNodeToClient ccfg =
+    -- Note: Ideally we'd just use `rt` to test Ledger config, but that would
+    -- require an Eq and Show instance for all ledger config types which we'd
+    -- like to avoid.
+    , testProperty ("roundtrip (comparing encoding) LedgerConfig") $ forAllShow
```
Would you use the `Blind` QC modifier instead of `forAllShow`? `forAllShow` does now shrinking :/

DavidEichmann replied:
Oh yes, good idea!

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Basics.hs:
```diff
@@ -141,10 +146,29 @@ data HardForkLedgerConfig xs = HardForkLedgerConfig {
 type instance LedgerCfg (LedgerState (HardForkBlock xs)) = HardForkLedgerConfig xs
 
+instance (CanHardFork xs, UpdateLedger (HardForkBlock xs))
+  => HasPartialLedgerConfig (HardForkBlock xs) where
```
Well, I made terrible progress here in the 24 days since I wrote that.

So I'll just ask you to help me: why is this instance now necessary? I generally think of `HasPartialLedgerConfig` as something an era's block type needs before you can put that era into the HFC. And we generally do not plan to nest HFC. So this instance smells wrong to me.

Why do we need it now?

DavidEichmann replied:
Ooops you're actually correct in that I can remove this. I think it is just a vestigial change from a previous revision where I was sending `PartialLedgerConfig` and hence required that all block types have a `HasPartialLedgerConfig` instance. Now we're sending `LedgerConfig` directly and no longer require this instance.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -1,29 +1,35 @@
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE QuantifiedConstraints      #-}
```
Is `QuantifiedConstraints` being used here? (Maybe via a type synonym? Because I'm not directly seeing it in diff here.)

Nor `OverloadedStrings`. So would you check this list for dead pragmas?

DavidEichmann replied:
Indeed I was able to revert these new language pragmas.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+  xs
+  = case version of
+    HardForkNodeToClientDisabled versionX -> case ccfgs of
+      (ccfg :: CodecConfig x) :* _ -> case xs of
```
Is this pattern signature necessary? I generally try to avoid those; they often do counter-intuitive binding. Please remove it, if you can.

DavidEichmann replied:
Will remove

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+
+-- | Encoding of `NP f xs` while filtering out the components for disabled eras
+-- as indicated by the `HardForkNodeToClientVersion xs`. disabled eras imply
+-- that the protocol version does not support those eras. Hence, omitting the
```
> disabled eras imply that the protocol version does not support those eras. Hence, omitting the corresponding elements is the correct behavior.

Are we assuming that the eras are in a sequence and that some prefix is enabled while the complementary suffix is disabled? We should `assert` that shape, if we are.

DavidEichmann replied:
That is indeed a property of the existing node-to-client versions and I expect the same from future versions. This function does **not** AFAICS make use of that assumption: we simple omit disabled eras regardless of position their. Hence I don't see a reason to add an assertion here. Am I missing something?

DavidEichmann replied:
In your mind what is an "expected arrangement"? Do you mean a run of 0 or more enabled followed by a run of 0 or more disabled? If we really want to enforce that property, then maybe we should change the `HardForkNodeToClientEnabled` constructor to store a `Telescope EraNodeToClientVersion EraNodeToClientVersion xs` instead of a `NP EraNodeToClientVersion xs` or something like that?

For now I've just added an assertion here and in `decodeNodeToClientNP`.

nfrisby replied:
Yep, that's what I meant. Interesting idea to use a `Telescope`.

I've opened https://github.com/input-output-hk/ouroboros-network/issues/3303

Thanks. Resolving this convo.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+                (\ccfg subVersionMay x -> K $ case subVersionMay of
+                  EraNodeToClientEnabled subVersion -> Just (encodeElement ccfg subVersion x)
+                  -- Omit disabled eras
+                  EraNodeToClientDisabled -> Nothing
```
Please align these `->` symbols.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+  (HardForkCodecConfig (PerEraCodecConfig ccfgs))
+  version
+  = case version of
+    HardForkNodeToClientDisabled versionX -> case ccfgs of
```
Please indent these alternatives (ie all lines between the `case version of` and `where` lines) two more spaces.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Summary.hs:
```diff
@@ -4,10 +4,12 @@
 {-# LANGUAGE DerivingStrategies         #-}
```
Probably don't need these `LANGUAGE` pragma changes?

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/Counting.hs:
```diff
@@ -1,20 +1,23 @@
-{-# LANGUAGE DataKinds           #-}
```
Probably don't need these `LANGUAGE` changes?

nfrisby reviewed (changes requested):
I've made another pass. There are many small things, a few medium things, and a couple high-level things. All of which I've asked you to carry the baton on.

Sorry for the delay. I wish I could get in to the details with you right now, but my brain doesn't have the throughput/context-switching capacity at the moment :(

DavidEichmann commented on ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs:
```diff
@@ -77,6 +86,82 @@ epochSlots = EpochSlots 100
+    <*> arbitrary
+    <*> arbitrary
+
+instance Arbitrary SafeSigner where
```
There is a `genSafeSigner` generator in `cardano-ledger-specs/byron/crypto/test/Test/Cardano/Crypto/Gen.hs`, but it leaves the passphrase blank rather than using an arbitrary pass phrase. According to @dcoutts, the passphrase "was security theatre. It added no value and considerably complexity". That said, we probably still want to generate some passphrase for the serialization tests, in which case, this instance is perhaps appropriate.

DavidEichmann commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Generators.hs:
```diff
@@ -152,6 +158,42 @@ instance ShelleyBasedEra era
+    <*> arbitrary
+    <*> arbitrary
+
+instance Monad m => Arbitrary (EpochInfo m) where
```
I've added explicit generator functions and only used them to generate a `ShelleyLedgerConfig` which is documented to have a fixed `EpochInfo`.

DavidEichmann commented on ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs:
```diff
@@ -637,3 +630,66 @@ instance c ~ MockCryptoCompatByron
+instance (CanMock (ShelleyEra c), CardanoHardForkConstraints c)
+      => Arbitrary (WithVersion (HardForkNodeToClientVersion (CardanoEras c)) (HardForkLedgerConfig (CardanoEras c))) where
+  arbitrary = WithVersion
+    -- Assume that all eras are enabled in the maximum supported version.
```
Correct

DavidEichmann commented on cabal.project:
```diff
@@ -186,8 +186,8 @@ source-repository-package
   location: https://github.com/input-output-hk/cardano-base
-  tag: 0f409343c3655c4bacd7fab385d392ec5d5cca98
-  --sha256: 0js76inb7avg8c39c9k2zsr77sycg2vadylgvsswdsba808p6hr9
+  tag: b6a215c42a28dc8b71b42946fe30256a333d34af
```
I've rebased.

DavidEichmann commented:
bors merge

nfrisby commented on ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs:
```diff
@@ -71,29 +73,77 @@ eraExamples =
      NP Examples (CardanoEras Crypto)
   -> Examples (CardanoBlock Crypto)
-combineEras = mconcat . hcollapse . hap eraInjections
+mkCardanoExamples perEraExamples = Golden.Examples {
```
Nice! Glad this worked out.

- Instead of the `::` annotations, do you think `@` type applications would be legible enough?

- The `coerce (foo (coerce . blah) ...)` seems off to me. I'm wondering if `coerce foo blah ...` would work instead. But I'm not confident one way or the other. Would you check eg `exampleApplyTxErr       = coerce (viaInject @WrapApplyTxErr) Golden.exampleApplyTxErr`?

DavidEichmann replied:
I originally had to shuffle the types around a lot to get things to work, but indeed I've managed to simplify a bit further as per your suggestions. Let me know if you're still not quite satisfied.

nfrisby replied:
Ah, OK, I see now that `viaInject` has a rank2 type, which spoils my idea to use a single `coerce`.

Since I got it into `ghci`, here's what I would suggest.

```
mkCardanoExamples perEraExamples = Golden.Examples {
        exampleBlock            = coerce $ viaInject @I                 (coerce Golden.exampleBlock)
      , exampleSerialisedBlock  =          viaInject                            Golden.exampleSerialisedBlock
      , exampleHeader           =          viaInject                            Golden.exampleHeader
      , exampleSerialisedHeader =          viaInject                            Golden.exampleSerialisedHeader
      , exampleHeaderHash       = coerce $ viaInject @WrapHeaderHash    (coerce Golden.exampleHeaderHash)
      , exampleGenTx            =          viaInject                            Golden.exampleGenTx
      , exampleGenTxId          = coerce $ viaInject @WrapGenTxId       (coerce Golden.exampleGenTxId)
      , exampleApplyTxErr       = coerce $ viaInject @WrapApplyTxErr    (coerce Golden.exampleApplyTxErr)
      , exampleQuery            =          viaInject                            Golden.exampleQuery
      , exampleResult           =          viaInject                            Golden.exampleResult
      , exampleAnnTip           =          viaInject                            Golden.exampleAnnTip
      , exampleLedgerState      =          viaInject                            Golden.exampleLedgerState
      , exampleChainDepState    = coerce $ viaInject @WrapChainDepState (coerce Golden.exampleChainDepState)
      , exampleExtLedgerState   =          viaInject                            Golden.exampleExtLedgerState

      , exampleLedgerConfig = exampleLedgerConfigCardano
      }
  where
    viaInject ::
         forall f. Inject f
      => (forall blk. Examples blk -> Labelled (f blk))
      -> Labelled (f (CardanoBlock Crypto))
```

Nothing else I could come up with was better than what you already have. My only grumble about what you have is that it makes the type of `viaInject` less legible.

I tried writing a bespoke `coerce'` that would handle the rank2 polymorphism of `viaInject` but it's confounded by the fact that some of those `coerce` occurrences above are working on multiple layers of application, so one combinator that abstracts over a single `f` doesn't cut it.

---

Feel free to adopt either what you have or my suggestion in this comment and click Resolve.

nfrisby commented on ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs:
```diff
@@ -71,29 +73,77 @@ eraExamples =
+            :* WrapPartialLedgerConfig (ShelleyPartialLedgerConfig lcAlonzo  TriggerHardForkNever)
+            :* Nil))
+        )
+      | lcByron   <- labelledLcByron
```
This looks like a cardinality explosion. I don't know if there's a sensible alternative, though.

DavidEichmann replied:
Humm I'll add an assertion on the resulting cardinality and error out with a message to fix the test... since this is just used in tests and the cardinality is fixed, I think a simple error is appropriate.

nfrisby replied:
Sounds reasonable.

- Only idea that might improve it: let's move the maximum size check into the definition of `Test.Util.Serialisation.Golden.goldenTests` and have it as a test failure (aborting the rest of that invocation) instead of an `error`.

- Also, let's write the check such that it doesn't compute the whole list length (which could be billions if something went wrong) and instead requires `null (drop maxSize examples)`.

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -19,15 +19,31 @@ module Test.Consensus.Shelley.Examples (
   ) where
 
+import qualified Data.Map as Map
+import           Data.Maybe (fromJust)
```
Please use `\case` etc or a local function definition instead of `fromJust`; the errors from that failing are entirely void of information.

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -126,15 +144,46 @@ fromShelleyLedgerExamples ShelleyLedgerExamples {
+
+    epochSize = EpochSize 4
+
+    slotLength' = secondsToNominalDiffTime 7
```
Style: inline this?

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -126,15 +144,46 @@ fromShelleyLedgerExamples ShelleyLedgerExamples {
+    slotLength' = secondsToNominalDiffTime 7
+    slotLength  = mkSlotLength slotLength'
+
+    genesis = testShelleyGenesis
```
Style: inline this?

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -126,15 +144,46 @@ fromShelleyLedgerExamples ShelleyLedgerExamples {
+  where
+    epochInfo = fixedEpochInfo epochSize slotLength
+
+    epochSize = EpochSize 4
```
Style: inline this?

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Generators.hs:
```diff
@@ -153,6 +158,45 @@ instance ShelleyBasedEra era
+  Generators for shelley ledger config
+-------------------------------------------------------------------------------}
+
+instance ( Mock (Crypto era)
```
Would you add a Haddock comment similar to that on `arbitraryGlobalsWithFixedEpochInfo` or referencing it, eg. Thanks.

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node/Serialisation.hs:
```diff
@@ -9,7 +15,7 @@ import           Control.Exception (Exception, throw)
 import qualified Data.ByteString.Lazy as Lazy
 import           Data.Typeable (Typeable)
 
-import           Cardano.Binary (fromCBOR, toCBOR)
```
GitHub says you Resolved this conversation, but I still see a diff here.

Is there more to it? (I've just now Unresolved it.)

DavidEichmann replied:
Will fix (for real this time).

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +210,124 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+    , encode @SlotLength slotLength
+    ]
+    where
+      unwrap = either
```
Style: please align these `=`

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +210,124 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+      => SerialiseNodeToClient (ShelleyBlock era) (ShelleyLedgerConfig era) where
+  decodeNodeToClient ccfg version = do
+    enforceSize "ShelleyLedgerConfig" 3
+    partialConfig <- decodeNodeToClient
```
Style: please consider aligning these `<-`

nfrisby commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:
```diff
@@ -198,16 +210,124 @@ instance ShelleyBasedEra era => HasPartialLedgerConfig (ShelleyBlock era) where
+        <> toCBOR @(SL.TranslationContext era) translationContext
+        <> encodeNodeToClient @(ShelleyBlock era) @TriggerHardFork ccfg version triggerHardFork
+
+instance ShelleyBasedEra era => SerialiseNodeToClientConstraints (ShelleyBlock era)
```
🙌

nfrisby commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Golden.hs:
```diff
@@ -194,6 +194,9 @@ goldenTests testName examples enc goldenFolder
       ]
   where
+    duplicateLabels :: [Maybe String]
+    duplicateLabels = labels \\ nub labels
```
I suggest using `Data.Set` here.

- It's not quadratic.

- This is always a surprise: `"aabbcc" \\ "abc" = "abc"`. It's fine here but it's not ideal.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Basics.hs:
```diff
@@ -7,12 +7,15 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
I see no other changes in this file; do we need to add these pragmas here?

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +48,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+-- as indicated by the `HardForkNodeToClientVersion xs`. Disabled eras imply
+-- that the protocol version does not support those eras. Hence, omitting the
+-- corresponding elements is the correct behavior.
+encodeNodeToClientNP ::
```
I think we should make `f` the first type variable, not `xs`.

Also, style: the first line layout should be

```
encodeNodeToClientNP :: forall f xs. SerialiseHFC xs
  => ...
```

DavidEichmann replied:
Doesn't the style guide suggest that we put the forall on a new line:

```haskell
fooBar ::
     forall a .. z. (Eq a, ..)
  => a
  -> ..
  -> ..
```

nfrisby replied:
Ah, egg on my face.

Yeah, follow the guide. Thanks for catching my mis-remembering.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +48,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+-- missing in the serialisation, then this throws an exception. In effect,
+-- deserialisation of product types is only supported when the sender uses an
+-- equal or superset of eras.
+decodeNodeToClientNP ::
```
Same as my comment on `encodeNodeToClientNP`: first line layout + `f` before `xs`

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs:
```diff
@@ -45,9 +48,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
 
 instance SerialiseHFC xs => SerialiseNodeToClientConstraints (HardForkBlock xs)
 
+instance SerialiseHFC xs => SerialiseNodeToClient (HardForkBlock xs) (HardForkLedgerConfig xs) where
```
[This same comment applies to all of the new instances in this module.]

Style: the `=` ends the first line, it doesn't start the second line. (See other instances that were already in this file, eg.)

nfrisby reviewed (changes requested):
Wonderful!

DavidEichmann commented on ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs:
```diff
@@ -71,29 +73,77 @@ eraExamples =
+          :* Nil
+          ) = hmap (WrapLedgerConfigList . fmap snd . Golden.exampleLedgerConfig) perEraExamples
+
+newtype WrapLedgerConfigList blk = WrapLedgerConfigList [LedgerConfig blk]
```
Thanks!

nfrisby reviewed (changes requested):
Only four Unresolved threads for you, all of which seem hopefully simple!

nfrisby commented on ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/ByronCompatibility.hs:
```diff
@@ -1,13 +1,15 @@
-{-# LANGUAGE TypeApplications      #-}
-{-# LANGUAGE TypeFamilies          #-}
-{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE DerivingStrategies         #-}
```
Some of these changes (at least `DerivingStrategies`) seem unnecessary.

DavidEichmann replied:
reverted

DavidEichmann commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Golden.hs:
```diff
@@ -172,19 +174,24 @@ goldenTestCBOR testName example enc goldenFile =
+  = testProperty testName $ counterexample
+        ("Too many examples (>" <> show maxSize <> "). Consider reducing the " <>
+        "number of examples so that tests don't take forever.")
+        False
```
@nfrisby Is there a more idiomatic way of making a test that always fails?

nfrisby replied:
No, this seems right to me.

The only alternative would be to make it a `.||.` property instead of using top-level guards. But I would lean towards the the current shape.

nfrisby commented on ouroboros-consensus-test/src/Test/Util/Serialisation/Golden.hs:
```diff
@@ -172,19 +174,24 @@ goldenTestCBOR testName example enc goldenFile =
     -- If there's just a single unlabelled example, no need for grouping,
     -- which makes the output more verbose.
   = goldenTestCBOR testName example enc (goldenFolder </> testName)
+  | not (null (drop maxSize examplesFull))
```
If this guard is the first, then you don't need to distinguish between `examples` and `examplesFull`.

DavidEichmann replied:
Good idea!

nfrisby commented:
I only see the two comments I just wrote as the Unresolved Conversations.

I sent a calendar invite for a short call at 5pm Paris time tomorrow (ie Wednesday). I'm hopeful maybe we can close this out then! (I can't see your calendar; I have an hour meeting at 6pm Paris time, but I'm otherwise available after 5pm Paris time. We can find another day if none of that works for you on short notice. Or maybe we can proceed async.)

nfrisby commented on ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:
```diff
@@ -61,9 +76,10 @@ codecConfig = ShelleyCodecConfig
 fromShelleyLedgerExamples
   :: ShelleyBasedEra era
-  => ShelleyLedgerExamples era
+  => Core.TranslationContext era
```
This parameter seems off to me. Two ideas:

- Maybe what we actually want is to add a `TranslationContext` upstream to the `ShelleyLedgerExamples` data type?

- In the meantime, we could use a new class here. It is not our intention to test with multiple `TranslationContexts`s -- but since the type varies per era, we can't trivially bake such a context into this definition. Hence this parameter. But since our intentions are all static and vary only wrt to era, we could use a class to emphasize that.

DavidEichmann replied:
@nfrisby Sorry for the delay. I totally agree with your first bullet I can make a PR for that in the cardano-ledger repo.

Your second bullet is simply an alternative to the current implementation, which can also be seen as a temporary solution before we can integrate upstream changes to cardano-ledger, correct? I don't really agree with the use of a class here; it seems over engineered. In particular, if I made a class it would only be used in this one module and only be used once per instance (i.e. per era), so it would be reduced to a complicated way of passing a function argument. We don't gain anything in terms of code reuse and even if did, this would all be soon replaced with the upstream changes. Is this convincing?

nfrisby replied:
Any progress with that upstream PR? Looks like nothing has changed in this part of the diff yet.

DavidEichmann replied:
@nfrisby I've made a PR here: https://github.com/input-output-hk/cardano-ledger/pull/2598

DavidEichmann replied:
In the latest [update](https://github.com/input-output-hk/ouroboros-network/compare/b3164abcf9d7790a66514af3390687ab00ce03eb..63095e6f38c282592f0413376fbde06941351fb9), I'm using the upstream example.

nfrisby reviewed (commented):
Hi David. I have just one comment on the new golden tests code.

Pawel relayed your question of whether we should split this PR. I'll need to look over the historical PR comments again to try to recall why the diff is so big. But at the moment I don't recall any lines of division along which it would 1) make sense and 2) be easy to split into multiple PRs. @DavidEichmann did you have any such lines of divisions in mind?

DavidEichmann commented:
In terms of breaking this PR down, it might be possible to do so by first introducing serialization instances, but they will in fact be unused until the later PRs are merged, which is not ideal. At least this pr is broken into smaller commits. I'm more wondering if it might be necessary in order to make Pawel's life easier (I dono if he's planning on reviewing this).

nfrisby commented on ouroboros-consensus-byronspec/src/Ouroboros/Consensus/ByronSpec/Ledger/Genesis.hs:
```diff
@@ -1,6 +1,12 @@
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE TypeApplications      #-}
+
+{-# OPTIONS_GHC -Wno-orphans  #-}
```
I don't see any orphans, at least not in the diff.

DavidEichmann replied:
removed

nfrisby commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:
```diff
@@ -1,21 +1,24 @@
+{-# LANGUAGE DisambiguateRecordFields   #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
I don't see any GND in this module. I wonder if more of these additional extensions are also dead-code now?

nfrisby commented:
I don't think it's worthwhile to break up the PR. Address my last comment and I'll do one final pass.

I've been reviewing this opportunistically between other more mainline Consensus work. I did a thorough review at least twice, but it's been a while. So there is a risk that something has slipped past me.

* GitHub's diff shows that only one file has changed since I last reviewed, and that's been the case for a while.

* There's one commit that has a lot of Haskell changes (30 files) but most are the reorganization of serialization instances. All of this PR's changes to the golden files are _new_ golden files -- so that evidences that the existing serialization was only refactored, not changed.

DavidEichmann commented:
I've rebased.

DavidEichmann commented:
Latest changes fix minor comments from above. I still need to integrate https://github.com/input-output-hk/cardano-ledger/pull/2598 once it's merged.

DavidEichmann commented:
Another rebase

DavidEichmann commented:
After speaking with @nfrisby, the main concern at the moment is conflicts with the work on UTxO HD which is high priority. I've tried a rough rebase on branch `davide/CAD-3877-utxo-hd-inmem_on-ledger-config-query` and the conflicts were not particularly bad but did require some manual resolution. We have decided to wait for UTxO HD to be merged before merging this PR.

dnadales commented:
@nfrisby if I understood @dcoutts correctly we'd like to merge this, are there pending comments or work to address?

nfrisby commented:
I don't think anything has changed since the previous explanation by David: https://github.com/input-output-hk/ouroboros-network/pull/3214#issuecomment-1042742868

Jasagredo commented:
What should we do with this PR? CC: @dnadales @nfrisby

dnadales commented:
> What should we do with this PR? CC: @dnadales @nfrisby

IIRC this PR is waiting for UTxO HD.
